### PR TITLE
Add tutorial for trainable tensor subclass

### PR DESCRIPTION
**Pull request by andrewor14**
**Summary:** The new tutorial provides an example of how to implement a trainable tensor subclass that wraps quantized data. This extends the existing `MyDTypeTensor` with a few necessary steps to ensure proper gradient updates, namely:

1. Define a differentiable constructor
2. Define backward pass for ops of interest (e.g. `torch.nn.functional.linear`)
3. Handle special ops used by the optimizer (e.g. `aten.add`, `aten.add_`)

**Test Plan:**
python tutorials/developer_api_guide/my_trainable_tensor_subclass.py

```
 * step 0: weight grad = tensor([-0.3854,  0.3851,  0.2130], device='cuda:0'), weight value = tensor([  41,  -73, -110], device='cuda:0', dtype=torch.int8)
 * step 1: weight grad = tensor([1.3488, 0.1335, 0.8074], device='cuda:0'), weight value = tensor([-107,  107,  116], device='cuda:0', dtype=torch.int8)
 * step 2: weight grad = tensor([-0.1505,  0.0958, -0.2342], device='cuda:0'), weight value = tensor([114, -98,  94], device='cuda:0', dtype=torch.int8)
 * step 3: weight grad = tensor([-0.4069,  0.7332,  0.1454], device='cuda:0'), weight value = tensor([-101,   54,  111], device='cuda:0', dtype=torch.int8)
 * step 4: weight grad = tensor([-0.0226,  0.0785, -0.0824], device='cuda:0'), weight value = tensor([  48,   31, -109], device='cuda:0', dtype=torch.int8)
 * step 5: weight grad = tensor([-2.0708,  0.6241, -2.0108], device='cuda:0'), weight value = tensor([112, -16,   5], device='cuda:0', dtype=torch.int8)
 * step 6: weight grad = tensor([-0.6014,  2.3908,  0.3403], device='cuda:0'), weight value = tensor([-17,  28,  69], device='cuda:0', dtype=torch.int8)
 * step 7: weight grad = tensor([ 1.1537, -2.1295,  0.5759], device='cuda:0'), weight value = tensor([-36, 116, 101], device='cuda:0', dtype=torch.int8)
 * step 8: weight grad = tensor([ 0.4791,  0.9322, -0.9088], device='cuda:0'), weight value = tensor([ 25, -24,  75], device='cuda:0', dtype=torch.int8)
 * step 9: weight grad = tensor([-0.4855,  0.1751, -0.4969], device='cuda:0'), weight value = tensor([ 26, 111,   1], device='cuda:0', dtype=torch.int8)
```

### CODE REVIEW

File: tutorials/developer_api_guide/my_trainable_tensor_subclass.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+    """
+    Handle the linear op with quantized weights.
+    For simplicity, we run both the forward and backward passes entirely in float.
+    """
+    assert isinstance(args[1], MyTrainableDTypeTensor)
+    if len(args) > 2 and args[2] is not None:
+        raise NotImplementedError("linear bias not yet supported")
+    return _QuantizedLinearOp.apply(args[0], args[1])
+
+@implements(aten.add_.Tensor)
```
Comment (jerryzh168) [inline tutorials/developer_api_guide/my_trainable_tensor_subclass.py:131]: are add_/add both going to be differentiable as well?

File: tutorials/developer_api_guide/my_trainable_tensor_subclass.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+
+
+########
+# Test #
+########
+
+class M(torch.nn.Module):
+    def __init__(self, *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+        self.linear = torch.nn.Linear(1024, 1024, bias=False)
```
Comment (jerryzh168) [inline tutorials/developer_api_guide/my_trainable_tensor_subclass.py:167]: nit: maybe changing to things like 512, 1024 so we can catch shape problems as well

File: tutorials/developer_api_guide/my_trainable_tensor_subclass.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+        ctx.save_for_backward(input_tensor, weight_tensor)
+        weight_tensor = weight_tensor.dequantize()
+        return torch.nn.functional.linear(input_tensor, weight_tensor)
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        input_tensor, weight_tensor = ctx.saved_tensors
+        weight_int = weight_tensor.layout_tensor.int_data
+        weight_scale = weight_tensor.layout_tensor.scale
+        grad_input = (grad_output * weight_scale) @ weight_int.to(grad_output.dtype)
```
Comment (jerryzh168) [inline tutorials/developer_api_guide/my_trainable_tensor_subclass.py:116]: can this call into F.linear impl (or some aten mm impl) of weight_tensor?

**Review by jerryzh168 — APPROVED** [review]
thanks, this looks great!

File: tutorials/developer_api_guide/my_trainable_tensor_subclass.py (hunk 1 above)
Comment (andrewor14) [reply tutorials/developer_api_guide/my_trainable_tensor_subclass.py:131]: I think they don't need to be. These are just used by the optimizer

**Comment by andrewor14** [comment]
Ok, merging this. Thanks!

File: tutorials/developer_api_guide/my_trainable_tensor_subclass.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+    ) -> torch.Tensor:
+        assert isinstance(weight_tensor, MyTrainableDTypeTensor)
+        ctx.save_for_backward(input_tensor, weight_tensor)
+        weight_tensor = weight_tensor.dequantize()
+        return torch.nn.functional.linear(input_tensor, weight_tensor)
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        input_tensor, weight_tensor = ctx.saved_tensors
+        grad_input = torch.matmul(grad_output, weight_tensor.dequantize())
```
Comment (jerryzh168) [inline tutorials/developer_api_guide/my_trainable_tensor_subclass.py:114]: sorry I guess I didn't meant to dequantize this, but was asking if this call into something like F.linear(grad_output, weight_tensor) and be dispatched to the quantized linear impl of weight_tensor.

Comment (andrewor14) [reply tutorials/developer_api_guide/my_trainable_tensor_subclass.py:114]: oh actually the old code was also doing dequantize math. I just rewrote it so it's cleaner. Not sure if I understand how we can dispatch to the quantized linear impl, since this is called from `F.linear(..., weight_subclass_tensor)` already?
